polynote-kernel/src/main/scala/polynote/kernel/dependency/DependencyFetcher.scala

```diff
@@ -33,157 +30,14 @@ trait DependencyFetcher[F[_]] {
-    val filteredArtifacts = allArtifacts.distinct.collect {
-      case (dependency, attributes, artifact) =>
-        s"${dependency.module}:${dependency.version}" -> artifact
-    }.distinct.map {
-      case (mv, artifact) => mv -> Cache.file[IO](artifact, logger = Some(logger)).leftMap(FileErrorException).run.flatMap(IO.fromEither)
-    }.toList
-
-    val downloadedFiles = downloaded.map {
-      case (url, ioF) =>
-        val ioWithUpdated = statusUpdates.publish1(UpdatedTasks(TaskInfo(url, s"Downloading $url", url, TaskStatus.Running) :: Nil)).bracket { _ =>
-          ioF
-        } { _ =>
-          statusUpdates.publish1(UpdatedTasks(TaskInfo(url, s"Downloading $url", url, TaskStatus.Complete) :: Nil))
-        }
-        "downloaded" -> ioWithUpdated
-    }
-
-    filteredArtifacts ++ downloadedFiles
-  }
-
-  private def resolveDependencies(
-    resolution: Resolution,
-    fetch: Fetch.Metadata[IO],
-    taskInfo: TaskInfo,
-    statusUpdates: Publish[IO, KernelStatusUpdate],
-    maxIterations: Int = 100
-  ): IO[Resolution] = {
-    // check whether we care about a missing resolution
-    def shouldErrorIfMissing(mv: (Module, String)): Boolean = {
-      resolution.rootDependencies.exists(dep => dep.module == mv._1 && dep.version == mv._2)
-    }
-
-    // reimplements ResolutionProcess.run, so we can update the iteration progress
-    def run(resolutionProcess: ResolutionProcess, iteration: Int): IO[Resolution] =
-      statusUpdates.publish1(UpdatedTasks(taskInfo.copy(progress = (iteration.toDouble / maxIterations * 255).toByte) :: Nil)) *> {
-        if (iteration > maxIterations) {
-          IO.pure(resolutionProcess.current)
-        } else {
-          resolutionProcess match {
-            case Done(res) if res.errorCache.keys.exists(shouldErrorIfMissing) =>
-              res.errorCache.map {
-                case (mv, err) if shouldErrorIfMissing(mv) =>
-                  val depStr = s"${mv._1}:${mv._2}"
-                  statusUpdates.publish1(
-                    UpdatedTasks(TaskInfo(
-                      id = s"${taskInfo.id}_$depStr",
-                      label = s"Error fetching dependency $depStr",
-                      detail = err.mkString("\n\n"),
-                      status = TaskStatus.Error
-                    ) :: taskInfo.copy(status = TaskStatus.Complete) :: Nil)
-                  )
-                case _ => IO.unit
-              }.toList.sequence *> IO.raiseError(new Exception("Dependency Resolution Error"))
-            case Done(res) =>
-              IO.pure(res)
-            case missing0 @ Missing(missing, _, _) =>
-              CoursierFetcher.fetchAll(missing, fetch).flatMap {
-                result => run(missing0.next(result), iteration + 1)
-              }
-            case cont @ Continue(_, _) =>
-              run(cont.nextNoCont, iteration + 1)
-          }
-        }
-    }
-
-    run(resolution.process, 0)
-  }
+trait URLDependencyFetcher extends DependencyFetcher[IO] {
```

Comment:
What I'm doing here is
1. Extract CoursierFetcher to its own file
2. Pull out the bits that are responsible for fetching URLs into a trait, since they're not specific to Coursier